### PR TITLE
fix: replace first and last name with single field

### LINKED ISSUE #4841: [BUG] Immich has separate first and last name fields and both are mandatory
### The bug

When creating or editing a user account, you are _required_ to provide both a first and last name. Failure to do so will not let you complete the user creation flow.  If you only have one name, or a name that doesn't readily decompose into two parts, or if those parts don't map cleanly to "first" and "last", you are out of luck.

I'm not sure if this is a design bug (Immich's data model assumes that people have names that come in at least two parts called "first" and "last") or an implementation bug (Immich's data model permits freeform names but the UI is incorrectly wired up with two name fields and both are marked mandatory).

Ideally, the name field should be a single free text "display name" field. If Immich is commited to binomial names, at least one of the fields should be optional.

### The OS that Immich Server is running on

docker on nixos

### Version of Immich Server

v1.84.0

### Version of Immich Mobile App

N/A

### Platform with the issue

- [X] Server
- [X] Web
- [ ] Mobile

### Your docker-compose.yml content

I'm not using docker-compose.yml, but a nixos config equivalent to [the official docker-compose](https://github.com/immich-app/immich/blob/main/docker/docker-compose.yml) but with redis and postgres running on the host.

### Your .env content

```nix
environment = {
    DB_HOSTNAME = "host.docker.inernal";
    DB_USERNAME = [redacted];
    DB_PASSWORD = [redacted];
    DB_DATABASE_NAME = "immich";

    REDIS_HOSTNAME = "host.docker.internal";
    REDIS_PASSWORD = [redacted];

    TYPESENSE_ENABLED = "true";
    TYPESENSE_HOST = "immich-typesense";
    TYPESENSE_API_KEY = [redacted];
    TYPESENSE_DATA_DIR = "/data";

    IMMICH_WEB_URL = "http://immich-web:3000";;
    IMMICH_SERVER_URL = "http://immich-server:3001";
    IMMICH_MACHINE_LEARNING_URL = "http://immich-ml:3003";
  };
```

### Reproduction steps

1. As an admin, click "add user".
2. Fill in the email and password but leave one (or both) of the name fields blank.
3. Click "create".

### Additional information

Setting the last name to a single space seems to appease it and has not _yet_ visibly broken anything.

[Falsehoods Programmers Believe about Names](https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/) is 13 years old at this point and still as relevant as ever.

Google dug themselves into a similar hole and it took them years to climb out of it, with ugly workarounds in the meantime like "put your actual name in the First Name field and put `.` in the Last Name field and it won't render the `.`, probably". But Google also had a lot more technical debt than Immich, so hopefully this won't take years to fix.

### PR DESCRIPTION
**Pull request by brianjaustin**
Resolves #4841 by removing the `firstName` and `lastName` user attributes, which are replaced with `fullName`.

I've done some local manual testing, but not gone over every path where changes were necessary. Please let me know if there are specific areas of concern that could do with more scrutiny.

# PR Checklist
## Web Checks
- [X] npm run lint (linting via ESLint)
- [X] npm run format (formatting via Prettier)
- [X] npm run check (Type checking via SvelteKit)
- [X] npm test (Tests via Jest)

## Server Checks
- [X] npm run lint (linting via ESLint)
- [X] npm run format (formatting via Prettier)
- [X] npm run check (Type checking via tsc)
- [X] npm test (Tests via Jest)

### CODE REVIEW

**Comment by jrasm91** [comment]
I would probably prefer to just use `name`. Is there a reason you went with full name?

File: server/src/infra/migrations/1699322864544-UserFullName.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { MigrationInterface, QueryRunner } from "typeorm";
+
+export class AddUserFullName1699322864544 implements MigrationInterface {
+    name = 'AddUserFullName1699322864544'
+
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        await queryRunner.query(`ALTER TABLE "users" ADD "fullName" character varying NOT NULL DEFAULT ''`);
+        await queryRunner.query(`UPDATE "users" SET "fullName" = CONCAT(COALESCE("firstName", ''), '', COALESCE("lastName", ''))`)
```
Comment (bo0tzz) [inline server/src/infra/migrations/1699322864544-UserFullName.ts:8]:
```suggestion
        await queryRunner.query(`UPDATE "users" SET "fullName" = CONCAT(COALESCE("firstName", ''), ' ', COALESCE("lastName", ''))`)
```

**Comment by brianjaustin** [comment]
> I would probably prefer to just use `name`. Is there a reason you went with full name?

@jrasm91 nothing big, mostly that HTML (and possibly other things?) uses `name` for something specific and I didn't want to accidentally cause confusion. I'm happy to change it to `name` if that's what you prefer

**Comment by jrasm91** [comment]
I would prefer that, thanks!

File: server/src/domain/auth/auth.service.ts [hunk 2]
```diff
@@ -274,8 +273,7 @@ export class AuthService {
       }
 
       user = await this.userCore.createUser({
-        firstName: profile.given_name || '',
-        lastName: profile.family_name || '',
+        name: `${profile.given_name || ''} ${profile.family_name || ''}`,
```
Comment (jrasm91) [inline server/src/domain/auth/auth.service.ts:276]: I think there is a name on the profile or full name. We should use that first and then fall back to what your have here.

File: server/src/infra/migrations/1699322864544-UserNameConsolidation.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        await queryRunner.query(`ALTER TABLE "users" ADD "name" character varying NOT NULL DEFAULT ''`);
+        await queryRunner.query(`UPDATE "users" SET "name" = CONCAT(COALESCE("firstName", ''), ' ', COALESCE("lastName", ''))`);
+        await queryRunner.query(`ALTER TABLE "users" DROP COLUMN "firstName"`);
+        await queryRunner.query(`ALTER TABLE "users" DROP COLUMN "lastName"`);
+    }
+
+    public async down(queryRunner: QueryRunner): Promise<void> {
+        await queryRunner.query(`ALTER TABLE "users" DROP COLUMN "name"`);
+        await queryRunner.query(`ALTER TABLE "users" ADD "lastName" character varying NOT NULL DEFAULT ''`);
```
Comment (jrasm91) [inline server/src/infra/migrations/1699322864544-UserNameConsolidation.ts:15]: Maybe default these to the email or something?

**Review by jrasm91 — COMMENTED** [review]
Looks really good. That is a lot of lines changed lol.

I don't know what is required as far as translations go. English for sure, but I think the rest might need to be made in the translation tool/web app we use for it.

**Comment by brianjaustin** [comment]
@jrasm91 I can definitely revert the changes to non-English locales if that's what works better with the translation process.

One other thing: I ran into an error when running the mobile tests. git-bisect says the change from `fullName` to `name` is responsible, but I'm not sure how exactly
<details>
<summary>flutter test error</summary>

```
00:04 +11 -1: /home/brian/projects/immich-app/immich/mobile/test/sync_service_test.dart: Test SyncService grouped (setUpAll) [E]
  RangeError (index): Index out of range: index should be less than 87: 65872
  dart:typed_data                                           Uint8List.[]
  package:isar/src/native/isar_reader_impl.dart 216:19      IsarReaderImpl._readUint24
  package:isar/src/native/isar_reader_impl.dart 239:20      IsarReaderImpl.readStringOrNull
  package:isar/src/native/isar_reader_impl.dart 224:12      IsarReaderImpl.readString
  package:immich_mobile/shared/models/user.g.dart 148:16    _userDeserialize
  package:isar/src/native/isar_collection_impl.dart 42:38   IsarCollectionImpl.deserializeObject
  package:isar/src/native/isar_collection_impl.dart 55:14   IsarCollectionImpl.deserializeObjectOrNull
  package:isar/src/native/isar_collection_impl.dart 182:22  IsarCollectionImpl.getAllSync.<fn>
  package:isar/src/common/isar_common.dart 170:56           IsarCommon.getTxnSync.<fn>
  package:isar/src/common/isar_common.dart 133:24           IsarCommon._beginTxnSync
  package:isar/src/common/isar_common.dart 170:14           IsarCommon.getTxnSync
  package:isar/src/native/isar_collection_impl.dart 174:17  IsarCollectionImpl.getAllSync
  package:isar/src/isar_collection.dart 30:12               IsarCollection.getSync
  package:immich_mobile/shared/models/store.dart 199:28     _getUser
  package:immich_mobile/shared/models/store.dart 102:30     StoreValue._extract
  package:immich_mobile/shared/models/store.dart 74:19      Store._onChangeListener
  ===== asynchronous gap ===========================
  dart:async                                                _StreamImpl.listen
  package:immich_mobile/shared/models/store.dart 19:45      Store.init
  test/sync_service_test.dart 74:13                         main.<fn>.<fn>
  ===== asynchronous gap ===========================
  dart:async                                                _CustomZone.registerUnaryCallback
  test/sync_service_test.dart 70:7                          main.<fn>.<fn>
  
  This test failed after it had already completed.
  Make sure to use a matching library which informs the test runner
  of pending async work.
  dart:typed_data                                           Uint8List.[]
  package:isar/src/native/isar_reader_impl.dart 216:19      IsarReaderImpl._readUint24
  package:isar/src/native/isar_reader_impl.dart 239:20      IsarReaderImpl.readStringOrNull
  package:isar/src/native/isar_reader_impl.dart 224:12      IsarReaderImpl.readString
  package:immich_mobile/shared/models/user.g.dart 148:16    _userDeserialize
  package:isar/src/native/isar_collection_impl.dart 42:38   IsarCollectionImpl.deserializeObject
  package:isar/src/native/isar_collection_impl.dart 55:14   IsarCollectionImpl.deserializeObjectOrNull
  package:isar/src/native/isar_collection_impl.dart 182:22  IsarCollectionImpl.getAllSync.<fn>
  package:isar/src/common/isar_common.dart 170:56           IsarCommon.getTxnSync.<fn>
  package:isar/src/common/isar_common.dart 133:24           IsarCommon._beginTxnSync
  package:isar/src/common/isar_common.dart 170:14           IsarCommon.getTxnSync
  package:isar/src/native/isar_collection_impl.dart 174:17  IsarCollectionImpl.getAllSync
  package:isar/src/isar_collection.dart 30:12               IsarCollection.getSync
  package:immich_mobile/shared/models/store.dart 199:28     _getUser
  package:immich_mobile/shared/models/store.dart 102:30     StoreValue._extract
  package:immich_mobile/shared/models/store.dart 74:19      Store._onChangeListener
  ===== asynchronous gap ===========================
  dart:async                                                _StreamImpl.listen
  package:immich_mobile/shared/models/store.dart 19:45      Store.init
  test/sync_service_test.dart 74:13                         main.<fn>.<fn>
  ===== asynchronous gap ===========================
  dart:async                                                _CustomZone.registerUnaryCallback
  test/sync_service_test.dart 70:7                          main.<fn>.<fn>
  
00:04 +12 -1: /home/brian/projects/immich-app/immich/mobile/test/sync_service_test.dart: Test SyncService grouped (setUpAll)
[INFO] [2023-11-09 20:09:40.557710] Upserted 3 assets into the DB
00:04 +13 -1: /home/brian/projects/immich-app/immich/mobile/test/sync_service_test.dart: Test SyncService grouped (setUpAll)
[INFO] [2023-11-09 20:09:40.570742] Upserted 3 assets into the DB
[INFO] [2023-11-09 20:09:40.580532] Upserted 2 assets into the DB
00:04 +14 -1: /home/brian/projects/immich-app/immich/mobile/test/sync_service_test.dart: Test SyncService grouped (setUpAll)
[INFO] [2023-11-09 20:09:40.593966] Upserted 3 assets into the DB
00:04 +17 -1: Some tests failed.     
```
</details>

**Comment by jrasm91** [comment]
Probably some upgrade script is needed for the mobile database

**Comment by alextran1502** [comment]
Hello, thank you for the PR, can you help remove all the translation files except [mobile/assets/i18n/en-US.json](https://github.com/immich-app/immich/pull/4915/files#diff-8242bf9941e87ec9d473bb52d2af2d174b4b393bd07056298a74a824dc98c40c)

**Review by alextran1502 — APPROVED** [review]

**Review by jrasm91 — APPROVED** [review]

**Comment by fyfrey** [comment]
This PR breaks the app when upgrading because no migration of the users in the database is performed.

**Comment by waclaw66** [comment]
This PR causes crashes because of `name[0]`, `name` is `string` not `array`.

**Comment by brianjaustin** [comment]
@waclaw66 which platform does that occur on?

**Comment by waclaw66** [comment]
Web

```
Cannot read properties of undefined (reading '0') - 500
TypeError: Cannot read properties of undefined (reading '0')
    at /src/immich/web/src/lib/components/shared-components/user-avatar.svelte:84:18
    at Object.$$render (/src/immich/web/node_modules/svelte/src/runtime/internal/ssr.js:156:16)
    at eval (/src/immich/web/src/lib/components/shared-components/navigation-bar/navigation-bar.svelte:105:123)
    at Object.$$render (/src/immich/web/node_modules/svelte/src/runtime/internal/ssr.js:156:16)
    at eval (/src/immich/web/src/lib/components/layouts/user-page-layout.svelte:41:96)
    at Object.$$render (/src/immich/web/node_modules/svelte/src/runtime/internal/ssr.js:156:16)
    at eval (/src/immich/web/src/routes/(user)/photos/+page.svelte:151:100)
    at Object.$$render (/src/immich/web/node_modules/svelte/src/runtime/internal/ssr.js:156:16)
    at Object.default (/src/immich/web/.svelte-kit/generated/root.svelte:44:41)
    at eval (/src/immich/web/src/routes/+layout.svelte:119:39)
```

**Comment by brianjaustin** [comment]
Hmm I wasn't able to replicate that locally (either with a new or migrated version). That error looks like for some reason `name` wasn't set, but I don't think that shouldn't be the case since the column is required

**Comment by jrasm91** [comment]
There was an existing issue for mounting preexisting installs that got fixed in the partner sharing PR.

**Pull request merged**